File: YukariLauncher/src/main/java/net/kdt/pojavlaunch/customcontrols/mouse/Scroller.java
package net.kdt.pojavlaunch.customcontrols.mouse;

import org.lwjgl.glfw.CallbackBridge;

public class Scroller {

    private float mScrollOvershootH, mScrollOvershootV;
    private final float mScrollThreshold;
/**
 * コンストラクタ。
 * このクラスの新しいインスタンスを初期化します。
 */
    public Scroller(float mScrollThreshold) {
        this.mScrollThreshold = mScrollThreshold;
    }

    /**
     * スクロールジェスチャーを実行します。
     * @param dx プライマリポインターのベクターX座標
     * @param dy プライマリポインターのベクターY座標
     */
    public void performScroll(float dx, float dy) {
        float hScroll = (dx / mScrollThreshold) + mScrollOvershootH;
        float vScroll = (dy / mScrollThreshold) + mScrollOvershootV;
        int hScrollRound = (int) hScroll, vScrollRound = (int) vScroll;
        if(hScrollRound != 0 || vScrollRound != 0) CallbackBridge.sendScroll(hScroll, vScroll);
        mScrollOvershootH = hScroll - hScrollRound;
        mScrollOvershootV = vScroll - vScrollRound;
    }

    /**
     * スクロールジェスチャーを実行します。
     * @param vector プライマリポインターの相対位置を格納する2成分ベクター
     */
    public void performScroll(float[] vector) {
        performScroll(vector[0], vector[1]);
    }

    /**
     * スクロールオーバーシュート値をリセットします。スクロールオーバーシュートによりスクロールが
     * ぎこちなくなりますが、スクロールジェスチャーの終了時にリセットしないと異常が発生します。
     */
    public void resetScrollOvershoot() {
        mScrollOvershootH = mScrollOvershootV = 0f;
    }
}

File: YukariLauncher/src/main/java/com/arata/yukarilauncher/ui/subassembly/customcontrols/mouse/Scroller.kt
package com.arata.yukarilauncher.ui.subassembly.customcontrols.mouse

import org.lwjgl.glfw.CallbackBridge

/** スクロール入力を処理するクラス。閾値を超えた移動をスクロールイベントに変換します。 */
class Scroller(private val mScrollThreshold: Float) {
    private var mScrollOvershootH = 0f
    private var mScrollOvershootV = 0f

    /**
     * スクロールを実行します。
     * @param dx X方向の移動量
     * @param dy Y方向の移動量
     */
    fun performScroll(dx: Float, dy: Float) {
        val hScroll = dx / mScrollThreshold + mScrollOvershootH
        val vScroll = dy / mScrollThreshold + mScrollOvershootV
        val hScrollRound = hScroll.toInt()
        val vScrollRound = vScroll.toInt()
        if (hScrollRound != 0 || vScrollRound != 0) CallbackBridge.sendScroll(hScroll.toDouble(), vScroll.toDouble())
        mScrollOvershootH = hScroll - hScrollRound
        mScrollOvershootV = vScroll - vScrollRound
    }

    /**
     * ベクトル配列からスクロールを実行します。
     * @param vector [x, y]の移動ベクトル
     */
    fun performScroll(vector: FloatArray) {
        performScroll(vector[0], vector[1])
    }

    /** スクロールのオーバーシュート値をリセットします。 */
    fun resetScrollOvershoot() {
        mScrollOvershootH = 0f
        mScrollOvershootV = 0f
    }
}